ast/parser_test.go

```diff
@@ -1302,6 +1333,194 @@ func TestFutureImportsExtraction(t *testing.T) {
+			module: `package test
+import future.strict
+p := "foo" if {
+	input.x == 1
+} else := "bar" if {
+	input.x == 2
+} else := "baz" if input.x == 3
+else := "qux"`,
+		},
+		{
+			note: "`if` keyword not used before else body",
+			module: `package test
+import future.strict
+p if {
+	input.x == 1
+} else {
+	input.x == 2
+}`,
+			expectedErrors: []string{"rego_parse_error: `if` keyword is required before rule body"},
+		},
+		{
+			note: "`if` keyword not used before else body (value assignment)",
+			module: `package test
+import future.strict
+p := "foo" if {
+	input.x == 1
+} else := "bar" {
+	input.x == 2
+}`,
+			expectedErrors: []string{"rego_parse_error: `if` keyword is required before rule body"},
+		},
+		{
+			note: "`contains` keyword used on partial set rule (const key)",
+			module: `package test
+import future.strict
+p contains "q"`,
+		},
+		{
+			note: "`contains` keyword used on partial set rule (ref-head, const key)",
+			module: `package test
+import future.strict
+p.q contains "r"`,
+		},
+		{
+			note: "`contains` keyword not used on partial set rule (const key)",
+			module: `package test
+import future.strict
+p.q`,
+			expectedErrors: []string{"rego_parse_error: `contains` keyword is required for partial set rules"},
+		},
+		{
+			note: "object definition (ref-head with implicit `true` value",
+			module: `package test
+import future.strict
+p.q.r`,
+		},
+		{
+			note: "`contains` keyword used on partial set rule (var key, no body)",
+			module: `package test
+import future.strict
+p contains input.x`,
+		},
+		{
+			note: "`contains` keyword not used on partial set rule (var key, no body)",
+			module: `package test
+import future.strict
+p[input.x]`,
+			expectedErrors: []string{"rego_parse_error: `contains` keyword is required for partial set rules"},
+		},
+		{
+			note: "`if` keyword not used on partial object rule (ref-head, var key, implicit `true` value, no body)",
+			module: `package test
+import future.strict
+p.q[input.x]`,
+		},
+		{
+			note: "`contains` keyword used on partial set rule (var key)",
+			module: `package test
+import future.strict
+p contains x if { x = input.x}`,
+		},
+		{
+			note: "`if` keyword used on partial map rule (would be multi-value without `if`)",
+			module: `package test
+import future.strict
+p[x] if { x = input.x}`,
+		},
+		{
+			note: "`contains` and `if` keyword not used on partial rule",
+			module: `package test
+import future.strict
+p[x] { x = input.x}`,
+			// The developer likely intended a partial set.
+			expectedErrors: []string{
+				"rego_parse_error: `contains` keyword is required for partial set rules",
+				"rego_parse_error: `if` keyword is required before rule body",
+			},
+		},
+		{
+			note: "`contains` and `if` keyword not used on partial object rule (ref-head)",
+			module: `package test
+import future.strict
+p.q[x] { x = input.x}`,
+			expectedErrors: []string{
+				"rego_parse_error: `if` keyword is required before rule body",
```

Comment:
For partial set rules, we get 2 errors for the `contains` and `if`. Here we get one. We would need `contains` as well, correct?

Reply:
No, this is a partial object rule that has a single-value leaf, not a multi-value set, so a `contains` error isn't expected. The test case `note` is a bit confusing, though (probably a copy/paste mistake); updating.

This is one of the reasons for why we're enforcing the use of `if` and `contains` in `1.0`; today, the policy

```rego
package play

a[b] { b := 1}

x.y[z] { z := 2 }
```

will produce: 

```json
{
    "a": [
        1
    ],
    "x": {
        "y": {
            "2": true
        }
    }
}
```

For the uninitiated, it seems very random why `a` would be a set, and `x.y` would be an object.

All rules declared with `if` is an object unless `contains` is also used, though:

```rego
package play
import future.keywords

a[b] if { b := 1}

x.y[z] if { z := 2 }

```
-->
```json
{
    "a": {
        "1": true
    },
    "x": {
        "y": {
            "2": true
        }
    }
}
```
&
```rego
package play
import future.keywords

a contains b if { b := 1}

x.y contains z if { z := 2 }

```
-->
```json
{
    "a": [
        1
    ],
    "x": {
        "y": [
            2
        ]
    }
}
```
As illustrated above, enforcing the use of `if` and `contains` removes the current ambiguity.